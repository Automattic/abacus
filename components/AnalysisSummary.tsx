import { Paper, Table, TableBody, TableCell, TableContainer, TableHead, TableRow } from '@material-ui/core'
import { format } from 'date-fns'
import _ from 'lodash'
import React, { useMemo } from 'react'

import {
  Analysis,
  AnalysisStrategyToHuman,
  ExperimentFull,
  MetricBare,
  Recommendation,
  RecommendationWarningToHuman,
  Variation,
} from '@/models'

function RecommendationString({
  recommendation,
  experiment,
}: {
  recommendation: Recommendation
  experiment: ExperimentFull
}) {
  if (recommendation.endExperiment) {
    if (recommendation.chosenVariationId) {
      const chosenVariation = experiment.variations.find(
        (variation) => variation.variationId === recommendation.chosenVariationId,
      ) as Variation
      return (
        <>
          End experiment; deploy <code>{chosenVariation.name}</code>
        </>
      )
    }
    return <>End experiment; deploy either variation</>
  }
  return <>Keep running</>
}

function ParticipantCounts({
  experiment,
  latestPrimaryMetricAnalyses,
}: {
  experiment: ExperimentFull
  latestPrimaryMetricAnalyses: Analysis[]
}) {
  const sortedVariations = _.orderBy(experiment.variations, ['isPrimary', 'name'], ['desc', 'asc'])
  return (
    <TableContainer component={Paper}>
      <Table>
        <TableHead>
          <TableRow>
            <TableCell>Strategy</TableCell>
            <TableCell>Total</TableCell>
            {sortedVariations.map((variation) => (
              <TableCell key={variation.variationId}>
                <code>{variation.name}</code>
              </TableCell>
            ))}
          </TableRow>
        </TableHead>
        <TableBody>
          {latestPrimaryMetricAnalyses.map((analysis) => (
            <TableRow key={analysis.analysisStrategy}>
              <TableCell>{AnalysisStrategyToHuman[analysis.analysisStrategy]}</TableCell>
              <TableCell>{analysis.participantStats.total}</TableCell>
              {sortedVariations.map((variation) => (
                <TableCell key={variation.variationId}>
                  {analysis.participantStats[`variation_${variation.variationId}`] || 0}
                </TableCell>
              ))}
            </TableRow>
          ))}
        </TableBody>
      </Table>
    </TableContainer>
  )
}

function LatestResults({
  experiment,
  metrics,
  metricAssignmentIdToLatestAnalyses,
}: {
  experiment: ExperimentFull
  metrics: MetricBare[]
  metricAssignmentIdToLatestAnalyses: { [key: number]: Analysis[] }
}) {
  const metricsById = useMemo(() => _.zipObject(_.map(metrics, 'metricId'), metrics), [metrics])
  const metricAssignmentsById = _.zipObject(
    _.map(experiment.metricAssignments, 'metricAssignmentId') as number[],
    experiment.metricAssignments,
  )
  return (
    <>
      {Object.entries(metricAssignmentIdToLatestAnalyses).map(([metricAssignmentId, metricAnalyses]) => (
        <div key={metricAssignmentId}>
          <div>
            <strong>Metric: </strong>
            <code>{metricsById[metricAssignmentsById[metricAssignmentId].metricId].name}</code>
          </div>
          <div>
            <strong>Attribution window: </strong>
            {metricAssignmentsById[metricAssignmentId].attributionWindowSeconds / 3600} hours
          </div>
          <div>
            <strong>Last analyzed: </strong>
            {format(metricAnalyses[0].analysisDatetime, 'yyyy-MM-dd')}
          </div>
          <TableContainer component={Paper}>
            <Table>
              <TableHead>
                <TableRow>
                  <TableCell>Strategy</TableCell>
                  <TableCell>Participants (not final)</TableCell>
                  <TableCell>Difference interval</TableCell>
                  <TableCell>Recommendation</TableCell>
                  <TableCell>Warnings</TableCell>
                </TableRow>
              </TableHead>
              <TableBody>
                {metricAnalyses.map((analysis) => (
                  <TableRow key={`${metricAssignmentId}_${analysis.analysisStrategy}`}>
                    <TableCell>{AnalysisStrategyToHuman[analysis.analysisStrategy]}</TableCell>
                    <TableCell>
                      {analysis.participantStats.total} ({analysis.participantStats.not_final})
                    </TableCell>
                    {analysis.metricEstimates && analysis.recommendation ? (
                      <>
                        <TableCell>
                          [{_.round(analysis.metricEstimates.diff.bottom, 4)},{' '}
                          {_.round(analysis.metricEstimates.diff.top, 4)}]
                        </TableCell>
                        <TableCell>
                          <RecommendationString recommendation={analysis.recommendation} experiment={experiment} />
                        </TableCell>
                        <TableCell>
                          {analysis.recommendation.warnings.map((warning) => (
                            <div key={warning}>{RecommendationWarningToHuman[warning]}</div>
                          ))}
                        </TableCell>
                      </>
                    ) : (
                      <>
                        <TableCell>N/A</TableCell>
                        <TableCell>N/A</TableCell>
                        <TableCell>Not analyzed yet</TableCell>
                      </>
                    )}
                  </TableRow>
                ))}
              </TableBody>
            </Table>
          </TableContainer>
        </div>
      ))}
    </>
  )
}

/**
 * Main component for summarizing experiment analyses.
 */
export default function AnalysisSummary({
  analyses,
  experiment,
  metrics,
  debugMode,
}: {
  analyses: Analysis[]
  experiment: ExperimentFull
  metrics: MetricBare[]
  debugMode?: boolean
}) {
  const metricAssignmentIdToLatestAnalyses = useMemo(
    () =>
      _.mapValues(_.groupBy(analyses, 'metricAssignmentId'), (metricAnalyses) => {
        metricAnalyses = _.orderBy(metricAnalyses, ['analysisDatetime'], ['desc'])
        return _.sortBy(
          _.filter(metricAnalyses, ['analysisDatetime', metricAnalyses[0].analysisDatetime]),
          'analysisStrategy',
        )
      }),
    [analyses],
  )

  if (analyses.length === 0) {
    return <h2>No analyses yet for {experiment.name}.</h2>
  }

  return (
    <>
      <h2>Analysis summary</h2>
      <p>Found {analyses.length} analysis objects in total.</p>

      <div className='analysis-participant-counts'>
        <h3>Participant counts for the primary metric</h3>
        <ParticipantCounts
          experiment={experiment}
          latestPrimaryMetricAnalyses={
            metricAssignmentIdToLatestAnalyses[experiment.getPrimaryMetricAssignmentId() as number]
          }
        />
      </div>

      <div className='analysis-latest-results'>
        <h3>Latest results by metric</h3>
        <LatestResults
          experiment={experiment}
          metrics={metrics}
          metricAssignmentIdToLatestAnalyses={metricAssignmentIdToLatestAnalyses}
        />
      </div>

      {debugMode ? <pre>{JSON.stringify(analyses, null, 2)}</pre> : ''}
    </>
  )
}
