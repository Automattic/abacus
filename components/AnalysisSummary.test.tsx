import { render } from '@testing-library/react'
import React from 'react'

import AnalysisSummary from './AnalysisSummary'
import {
  Analysis,
  AnalysisStrategy,
  AttributionWindowSeconds,
  ExperimentFull,
  MetricAssignment,
  Platform,
  RecommendationReason,
  Status,
  Variation,
} from '@/models'

const experiment: ExperimentFull = new ExperimentFull({
  experimentId: 1,
  name: 'experiment_1',
  startDatetime: new Date(2020, 5, 4),
  endDatetime: new Date(2020, 6, 4),
  status: Status.Completed,
  platform: Platform.Calypso,
  ownerLogin: 'test_a11n',
  description: 'Experiment with things. Change stuff. Profit.',
  existingUsersAllowed: false,
  p2Url: 'https://wordpress.com/experiment_1',
  exposureEvents: null,
  variations: [
    new Variation({
      variationId: 2,
      name: 'test',
      isDefault: false,
      allocatedPercentage: 40,
    }),
    new Variation({
      variationId: 1,
      name: 'control',
      isDefault: true,
      allocatedPercentage: 60,
    }),
  ],
  metricAssignments: [
    new MetricAssignment({
      metricAssignmentId: 123,
      metricId: 1,
      experimentId: 1,
      attributionWindowSeconds: AttributionWindowSeconds.OneWeek,
      changeExpected: true,
      isPrimary: true,
      minDifference: 0.1,
    }),
  ],
  segmentAssignments: [],
})
const analyses: Analysis[] = [
  new Analysis({
    metricAssignmentId: 123,
    analysisStrategy: AnalysisStrategy.IttPure,
    participantStats: {
      total: 1000,
      not_final: 100,
      variation_1: 600,
      variation_2: 400,
    },
    metricEstimates: {
      diff: { estimate: 0.0, bottom: -0.01, top: 0.01 },
      variation_1: { estimate: 0.12, bottom: 0, top: 10.0 },
      variation_2: { estimate: -0.12, bottom: -1.123, top: 1.0 },
    },
    recommendation: {
      endExperiment: true,
      chosenVariationId: 2,
      reason: RecommendationReason.CiInRope,
      warnings: [],
    },
    analysisDatetime: new Date(2020, 4, 10),
  }),
  new Analysis({
    metricAssignmentId: 123,
    analysisStrategy: AnalysisStrategy.MittNoCrossovers,
    participantStats: {
      total: 900,
      not_final: 90,
      variation_1: 540,
      variation_2: 360,
    },
    metricEstimates: {
      diff: { estimate: 0.0, bottom: -0.01, top: 0.01 },
      variation_1: { estimate: 0.12, bottom: 0, top: 10.0 },
      variation_2: { estimate: -0.12, bottom: -1.123, top: 1.0 },
    },
    recommendation: {
      endExperiment: true,
      chosenVariationId: 2,
      reason: RecommendationReason.CiInRope,
      warnings: [],
    },
    analysisDatetime: new Date(2020, 4, 10),
  }),
  new Analysis({
    metricAssignmentId: 123,
    analysisStrategy: AnalysisStrategy.MittNoSpammers,
    participantStats: {
      total: 850,
      not_final: 85,
      variation_1: 510,
      variation_2: 340,
    },
    metricEstimates: {
      diff: { estimate: 0.0, bottom: -0.01, top: 0.01 },
      variation_1: { estimate: 0.12, bottom: 0, top: 10.0 },
      variation_2: { estimate: -0.12, bottom: -1.123, top: 1.0 },
    },
    recommendation: {
      endExperiment: true,
      chosenVariationId: 2,
      reason: RecommendationReason.CiInRope,
      warnings: [],
    },
    analysisDatetime: new Date(2020, 4, 10),
  }),
  new Analysis({
    metricAssignmentId: 123,
    analysisStrategy: AnalysisStrategy.MittNoSpammersNoCrossovers,
    participantStats: {
      total: 800,
      not_final: 80,
      variation_1: 480,
      variation_2: 320,
    },
    metricEstimates: {
      diff: { estimate: 0.0, bottom: -0.01, top: 0.01 },
      variation_1: { estimate: 0.12, bottom: 0, top: 10.0 },
      variation_2: { estimate: -0.12, bottom: -1.123, top: 1.0 },
    },
    recommendation: {
      endExperiment: true,
      chosenVariationId: 2,
      reason: RecommendationReason.CiInRope,
      warnings: [],
    },
    analysisDatetime: new Date(2020, 4, 10),
  }),
  new Analysis({
    metricAssignmentId: 123,
    analysisStrategy: AnalysisStrategy.PpNaive,
    participantStats: {
      total: 700,
      not_final: 70,
      variation_1: 420,
      variation_2: 280,
    },
    metricEstimates: {
      diff: { estimate: 0.0, bottom: -0.01, top: 0.01 },
      variation_1: { estimate: 0.12, bottom: 0, top: 10.0 },
      variation_2: { estimate: -0.12, bottom: -1.123, top: 1.0 },
    },
    recommendation: {
      endExperiment: true,
      chosenVariationId: 2,
      reason: RecommendationReason.CiInRope,
      warnings: [],
    },
    analysisDatetime: new Date(2020, 4, 10),
  }),
]

test('renders an appropriate message with no analyses', () => {
  const { container } = render(<AnalysisSummary analyses={[]} experiment={experiment} metrics={[]} />)
  expect(container).toHaveTextContent('No analyses yet for experiment_1.')
})

test('renders an appropriate message with some analyses', () => {
  const { container } = render(<AnalysisSummary analyses={analyses} experiment={experiment} metrics={[]} />)
  expect(container).toHaveTextContent('Found 5 analysis objects in total. There are 0 metrics in the system.')
})

test('shows the analyses JSON in debug mode', () => {
  const { container } = render(
    <AnalysisSummary analyses={analyses} experiment={experiment} metrics={[]} debugMode={true} />,
  )
  expect(container.querySelector('pre')).not.toBeNull()
})

// if (analyses.length === 0) {
//   return <h2>No analyses yet for {experiment.name}.</h2>
// }
//
// return (
//   <>
//     <h2>Analysis summary</h2>
//     <p>
//       Found {analyses.length} analysis objects in total. There are {metrics.length} metrics in the system.
//     </p>
//
//     {debugMode ? <pre>{JSON.stringify(analyses, null, 2)}</pre> : ''}
//   </>
// )
