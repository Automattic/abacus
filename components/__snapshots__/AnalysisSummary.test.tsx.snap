// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`shows the analyses JSON in debug mode 1`] = `
<pre>
  [
  {
    "metricAssignmentId": 123,
    "analysisStrategy": "itt_pure",
    "participantStats": {
      "total": 1000,
      "not_final": 100,
      "variation_1": 600,
      "variation_2": 400
    },
    "metricEstimates": {
      "diff": {
        "estimate": 0,
        "bottom": -0.01,
        "top": 0.01
      },
      "variation_1": {
        "estimate": 0.12,
        "bottom": 0,
        "top": 10
      },
      "variation_2": {
        "estimate": -0.12,
        "bottom": -1.123,
        "top": 1
      }
    },
    "recommendation": {
      "endExperiment": true,
      "chosenVariationId": 2,
      "reason": "ci_in_rope",
      "warnings": [
        "short_period",
        "wide_ci"
      ]
    },
    "analysisDatetime": "2020-05-10T00:00:00.000Z"
  },
  {
    "metricAssignmentId": 123,
    "analysisStrategy": "mitt_no_crossovers",
    "participantStats": {
      "total": 900,
      "not_final": 90,
      "variation_1": 540,
      "variation_2": 360
    },
    "metricEstimates": {
      "diff": {
        "estimate": 0,
        "bottom": -0.01,
        "top": 0.01
      },
      "variation_1": {
        "estimate": 0.12,
        "bottom": 0,
        "top": 10
      },
      "variation_2": {
        "estimate": -0.12,
        "bottom": -1.123,
        "top": 1
      }
    },
    "recommendation": {
      "endExperiment": false,
      "chosenVariationId": null,
      "reason": "rope_in_ci",
      "warnings": []
    },
    "analysisDatetime": "2020-05-10T00:00:00.000Z"
  },
  {
    "metricAssignmentId": 123,
    "analysisStrategy": "mitt_no_spammers",
    "participantStats": {
      "total": 850,
      "not_final": 85,
      "variation_1": 510,
      "variation_2": 340
    },
    "metricEstimates": {
      "diff": {
        "estimate": 0,
        "bottom": -0.01,
        "top": 0.01
      },
      "variation_1": {
        "estimate": 0.12,
        "bottom": 0,
        "top": 10
      },
      "variation_2": {
        "estimate": -0.12,
        "bottom": -1.123,
        "top": 1
      }
    },
    "recommendation": {
      "endExperiment": true,
      "chosenVariationId": null,
      "reason": "ci_in_rope",
      "warnings": []
    },
    "analysisDatetime": "2020-05-10T00:00:00.000Z"
  },
  {
    "metricAssignmentId": 123,
    "analysisStrategy": "mitt_no_spammers_no_crossovers",
    "participantStats": {
      "total": 800,
      "not_final": 80,
      "variation_1": 480,
      "variation_2": 320
    },
    "metricEstimates": {
      "diff": {
        "estimate": 0,
        "bottom": -0.01,
        "top": 0.01
      },
      "variation_1": {
        "estimate": 0.12,
        "bottom": 0,
        "top": 10
      },
      "variation_2": {
        "estimate": -0.12,
        "bottom": -1.123,
        "top": 1
      }
    },
    "recommendation": {
      "endExperiment": true,
      "chosenVariationId": 2,
      "reason": "ci_in_rope",
      "warnings": [
        "short_period",
        "wide_ci"
      ]
    },
    "analysisDatetime": "2020-05-10T00:00:00.000Z"
  },
  {
    "metricAssignmentId": 123,
    "analysisStrategy": "pp_naive",
    "participantStats": {
      "total": 700,
      "not_final": 70,
      "variation_1": 420,
      "variation_2": 280
    },
    "metricEstimates": {
      "diff": {
        "estimate": 0,
        "bottom": -0.01,
        "top": 0.01
      },
      "variation_1": {
        "estimate": 0.12,
        "bottom": 0,
        "top": 10
      },
      "variation_2": {
        "estimate": -0.12,
        "bottom": -1.123,
        "top": 1
      }
    },
    "recommendation": {
      "endExperiment": true,
      "chosenVariationId": 2,
      "reason": "ci_in_rope",
      "warnings": [
        "short_period",
        "wide_ci"
      ]
    },
    "analysisDatetime": "2020-05-10T00:00:00.000Z"
  },
  {
    "metricAssignmentId": 123,
    "analysisStrategy": "itt_pure",
    "participantStats": {
      "total": 100,
      "not_final": 10,
      "variation_1": 60,
      "variation_2": 40
    },
    "metricEstimates": {
      "diff": {
        "estimate": 0,
        "bottom": -0.01,
        "top": 0.01
      },
      "variation_1": {
        "estimate": 0.12,
        "bottom": 0,
        "top": 10
      },
      "variation_2": {
        "estimate": -0.12,
        "bottom": -1.123,
        "top": 1
      }
    },
    "recommendation": {
      "endExperiment": true,
      "chosenVariationId": 2,
      "reason": "ci_in_rope",
      "warnings": [
        "short_period",
        "wide_ci"
      ]
    },
    "analysisDatetime": "2020-05-09T00:00:00.000Z"
  },
  {
    "metricAssignmentId": 123,
    "analysisStrategy": "mitt_no_crossovers",
    "participantStats": {
      "total": 90,
      "not_final": 9,
      "variation_1": 54,
      "variation_2": 36
    },
    "metricEstimates": {
      "diff": {
        "estimate": 0,
        "bottom": -0.01,
        "top": 0.01
      },
      "variation_1": {
        "estimate": 0.12,
        "bottom": 0,
        "top": 10
      },
      "variation_2": {
        "estimate": -0.12,
        "bottom": -1.123,
        "top": 1
      }
    },
    "recommendation": {
      "endExperiment": false,
      "chosenVariationId": null,
      "reason": "rope_in_ci",
      "warnings": []
    },
    "analysisDatetime": "2020-05-09T00:00:00.000Z"
  },
  {
    "metricAssignmentId": 123,
    "analysisStrategy": "mitt_no_spammers",
    "participantStats": {
      "total": 85,
      "not_final": 8,
      "variation_1": 51,
      "variation_2": 34
    },
    "metricEstimates": {
      "diff": {
        "estimate": 0,
        "bottom": -0.01,
        "top": 0.01
      },
      "variation_1": {
        "estimate": 0.12,
        "bottom": 0,
        "top": 10
      },
      "variation_2": {
        "estimate": -0.12,
        "bottom": -1.123,
        "top": 1
      }
    },
    "recommendation": {
      "endExperiment": true,
      "chosenVariationId": null,
      "reason": "ci_in_rope",
      "warnings": []
    },
    "analysisDatetime": "2020-05-09T00:00:00.000Z"
  },
  {
    "metricAssignmentId": 123,
    "analysisStrategy": "mitt_no_spammers_no_crossovers",
    "participantStats": {
      "total": 80,
      "not_final": 8,
      "variation_1": 48,
      "variation_2": 32
    },
    "metricEstimates": {
      "diff": {
        "estimate": 0,
        "bottom": -0.01,
        "top": 0.01
      },
      "variation_1": {
        "estimate": 0.12,
        "bottom": 0,
        "top": 10
      },
      "variation_2": {
        "estimate": -0.12,
        "bottom": -1.123,
        "top": 1
      }
    },
    "recommendation": {
      "endExperiment": true,
      "chosenVariationId": 2,
      "reason": "ci_in_rope",
      "warnings": [
        "short_period",
        "wide_ci"
      ]
    },
    "analysisDatetime": "2020-05-09T00:00:00.000Z"
  },
  {
    "metricAssignmentId": 123,
    "analysisStrategy": "pp_naive",
    "participantStats": {
      "total": 70,
      "not_final": 7,
      "variation_1": 42,
      "variation_2": 28
    },
    "metricEstimates": {
      "diff": {
        "estimate": 0,
        "bottom": -0.01,
        "top": 0.01
      },
      "variation_1": {
        "estimate": 0.12,
        "bottom": 0,
        "top": 10
      },
      "variation_2": {
        "estimate": -0.12,
        "bottom": -1.123,
        "top": 1
      }
    },
    "recommendation": {
      "endExperiment": true,
      "chosenVariationId": 2,
      "reason": "ci_in_rope",
      "warnings": [
        "short_period",
        "wide_ci"
      ]
    },
    "analysisDatetime": "2020-05-09T00:00:00.000Z"
  },
  {
    "metricAssignmentId": 124,
    "analysisStrategy": "itt_pure",
    "participantStats": {
      "total": 10,
      "not_final": 10,
      "variation_1": 10
    },
    "metricEstimates": null,
    "recommendation": null,
    "analysisDatetime": "2020-05-10T00:00:00.000Z"
  }
]
</pre>
`;
